L1Trigger/TrackFindingTMTT/src/DigitalStub.cc

```diff
@@ -0,0 +1,457 @@
+        chosenRofPhi_(settings->chosenRofPhi()),
+
+        // Number of q/Pt bins in Hough  transform array.
+        nbinsPt_((int)settings->houghNbinsPt()) {
+    // Calculate multipliers to digitize the floating point numbers.
+    phiSMult_ = pow(2, phiSBits_) / phiSRange_;
+    rtMult_ = pow(2, rtBits_) / rtRange_;
+    zMult_ = pow(2, zBits_) / zRange_;
+    phiOMult_ = pow(2, phiOBits_) / phiORange_;
+
+    bendMult_ =
+        4.;  // No precision lost by digitization, since original bend (after encoding) has steps of 0.25 (in units of pitch).
+    bendRange_ = round(pow(2, bendBits_) / bendMult_);  // discrete values, so digitisation different
+  }
+
+  //=== Initialize stub with original, floating point stub coords,
+  //=== range of m bin (= q/Pt bin) values allowed by bend filter,
+  //=== normal & "reduced" tracker layer of stub, stub bend, and pitch & seperation of module,
+  //=== and half-length of strip or pixel in r and in z, and if it's in barrel, tilted barrel and/or PS module.
+
+  void DigitalStub::init(float phi_orig,
+                         float r_orig,
+                         float z_orig,
+                         unsigned int min_qOverPt_bin_orig,
+                         unsigned int max_qOverPt_bin_orig,
+                         unsigned int layerID,
+                         unsigned int layerIDreduced,
+                         float bend_orig,
+                         float pitch,
+                         float sep,
+                         float rErr,
+                         float zErr,
+                         bool barrel,
+                         bool tiltedBarrel,
+                         bool psModule) {
+    ranInit_ = true;  // Note we ran init().
+    // Variables in HT.
+    phi_orig_ = phi_orig;
+    r_orig_ = r_orig;
+    z_orig_ = z_orig;
+    min_qOverPt_bin_orig_ = min_qOverPt_bin_orig;
+    max_qOverPt_bin_orig_ = max_qOverPt_bin_orig;
+    layerID_ = layerID;
+    layerIDreduced_ = layerIDreduced;
+    // Variables exclusively in GP.
+    bend_orig_ = bend_orig;
+    rErr_orig_ = rErr;
+    zErr_orig_ = zErr;
+
+    // Calculate unique module type ID, allowing pitch/sep of module to be determined.
+    // (N.B. Module types 0 & 1 have identical pitch & sep, but one is tilted & one is flat barrel module).
+
+    moduleType_ = 999;
+    // EJC new (additional) module type for tilted geometry
+    const vector<float> pitchVsType = {0.0099, 0.0099, 0.0099, 0.0099, 0.0089, 0.0099, 0.0089, 0.0089};
+    const vector<float> sepVsType = {0.26, 0.26, 0.16, 0.4, 0.18, 0.4, 0.18, 0.4};
+    const vector<bool> barrelVsType = {true, true, true, true, true, false, false, false};
+    const vector<bool> psVsType = {true, true, true, true, false, true, false, false};
+    const vector<bool> tiltedVsType = {false, true, false, true, false, false, false, false};
+    if (pitchVsType.size() != sepVsType.size())
+      throw cms::Exception("DigitalStub: module type array size wrong");
```

Comment:
Since I prefer to keep the std::vector, I made them "static const", deleted the check on the size, and adjusting the vector initialization spacing to make it easy to check that they are the same size by reading the code.
Exception fixed.

Reply:
These parameters are now set in the python cfg, and processed by class TrackerModule instead of DigitalStub. This means that they can't be of type constexpr, since they are not known at compile time.